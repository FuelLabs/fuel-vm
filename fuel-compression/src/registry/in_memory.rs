use std::collections::HashMap;

use super::{
    db::*,
    key::RawKey,
    Key,
    Table,
};

/// Simple and inefficient in-memory registry for testing purposes.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct InMemoryRegistry {
    next_keys: HashMap<&'static str, RawKey>,
    storage: HashMap<&'static str, HashMap<RawKey, Vec<u8>>>,
    index: HashMap<&'static str, HashMap<Vec<u8>, RawKey>>,
}

impl RegistryDb for InMemoryRegistry {
    fn next_key<T: Table>(&self) -> Key<T> {
        Key::from_raw(self.next_keys.get(T::NAME).copied().unwrap_or(RawKey::ZERO))
    }

    fn read<T: Table>(&self, key: Key<T>) -> T::Type {
        if key == Key::DEFAULT_VALUE {
            return T::Type::default();
        }

        self.storage
            .get(T::NAME)
            .and_then(|table| table.get(&key.raw()))
            .map(|bytes| postcard::from_bytes(bytes).expect("Invalid value in registry"))
            .unwrap_or_default()
    }

    fn batch_write<T: Table>(&mut self, start_key: Key<T>, values: Vec<T::Type>) {
        let empty = values.is_empty();
        if !empty && start_key == Key::DEFAULT_VALUE {
            panic!("Cannot write to the default value key");
        }
        let table = self.storage.entry(T::NAME).or_default();
        let mut key = start_key.raw();
        for value in values.into_iter() {
            let value = postcard::to_stdvec(&value).unwrap();
            let mut prefix = value.clone();
            prefix.truncate(32);
            self.index.entry(T::NAME).or_default().insert(prefix, key);
            table.insert(key, value);
            key = key.next();
        }
        if !empty {
            self.next_keys.insert(T::NAME, key);
        }
    }

    fn index_lookup<T: Table>(&self, value: &T::Type) -> Option<Key<T>> {
        if *value == T::Type::default() {
            return Some(Key::DEFAULT_VALUE);
        }

        let needle = postcard::to_stdvec(value).unwrap();
        let mut prefix = needle.clone();
        prefix.truncate(32);
        if let Some(cand) = self.index.get(T::NAME)?.get(&prefix).copied() {
            let cand_val = self.storage.get(T::NAME)?.get(&cand)?;
            if *cand_val == needle {
                return Some(Key::from_raw(cand));
            }
        }

        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        tables,
        Key,
    };

    #[test]
    fn in_memory_registry_works() {
        let mut reg = InMemoryRegistry::default();

        // Empty
        assert_eq!(
            reg.read(Key::<tables::AssetId>::try_from(100).unwrap()),
            [0; 32]
        );

        // Write
        reg.batch_write(
            Key::<tables::AssetId>::from_raw(RawKey::try_from(100u32).unwrap()),
            vec![[1; 32], [2; 32]],
        );

        // Read
        assert_eq!(
            reg.read(Key::<tables::AssetId>::try_from(100).unwrap()),
            [1; 32]
        );

        // Index
        assert_eq!(
            reg.index_lookup(&[1; 32]),
            Some(Key::<tables::AssetId>::try_from(100).unwrap())
        );
    }
}
