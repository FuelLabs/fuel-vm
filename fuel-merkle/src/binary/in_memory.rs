use crate::common::{Bytes32, ProofSet};
use crate::{binary, common};

type StorageMap = common::StorageMap<binary::merkle_tree::NodesTable>;
type BinaryMerkleTree = binary::MerkleTree<StorageMap>;

pub struct MerkleTree {
    tree: BinaryMerkleTree,
}

impl MerkleTree {
    pub fn new() -> Self {
        Self {
            tree: BinaryMerkleTree::new(StorageMap::new()),
        }
    }

    pub fn push(&mut self, data: &[u8]) {
        let _ = self.tree.push(data);
    }

    pub fn root(&mut self) -> Bytes32 {
        self.tree.root().unwrap()
    }

    pub fn prove(&mut self, proof_index: u64) -> Option<(Bytes32, ProofSet)> {
        self.tree.prove(proof_index).ok()
    }
}

impl Default for MerkleTree {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use binary::{empty_sum, leaf_sum, node_sum};
    use fuel_merkle_test_helpers::TEST_DATA;

    #[test]
    fn root_returns_the_empty_root_for_0_leaves() {
        let mut tree = MerkleTree::new();

        let root = tree.root();
        assert_eq!(root, empty_sum().clone());
    }

    #[test]
    fn root_returns_the_merkle_root_for_1_leaf() {
        let mut tree = MerkleTree::new();

        let data = &TEST_DATA[0..1]; // 1 leaf
        for datum in data.iter() {
            tree.push(datum);
        }

        let leaf_0 = leaf_sum(data[0]);

        let root = tree.root();
        assert_eq!(root, leaf_0);
    }

    #[test]
    fn root_returns_the_merkle_root_for_7_leaves() {
        let mut tree = MerkleTree::new();

        let data = &TEST_DATA[0..7]; // 7 leaves
        for datum in data.iter() {
            tree.push(datum);
        }

        //               07
        //              /  \
        //             /    \
        //            /      \
        //           /        \
        //          /          \
        //         /            \
        //       03              11
        //      /  \            /  \
        //     /    \          /    \
        //   01      05      09      \
        //  /  \    /  \    /  \      \
        // 00  02  04  06  08  10     12
        // 00  01  02  03  04  05     06

        let leaf_0 = leaf_sum(data[0]);
        let leaf_1 = leaf_sum(data[1]);
        let leaf_2 = leaf_sum(data[2]);
        let leaf_3 = leaf_sum(data[3]);
        let leaf_4 = leaf_sum(data[4]);
        let leaf_5 = leaf_sum(data[5]);
        let leaf_6 = leaf_sum(data[6]);

        let node_1 = node_sum(&leaf_0, &leaf_1);
        let node_5 = node_sum(&leaf_2, &leaf_3);
        let node_3 = node_sum(&node_1, &node_5);
        let node_9 = node_sum(&leaf_4, &leaf_5);
        let node_11 = node_sum(&node_9, &leaf_6);
        let node_7 = node_sum(&node_3, &node_11);

        let root = tree.root();
        assert_eq!(root, node_7);
    }

    #[test]
    fn prove_returns_none_for_0_leaves() {
        let mut tree = MerkleTree::new();

        let proof = tree.prove(0);
        assert!(proof.is_none());
    }

    #[test]
    fn prove_returns_none_when_index_is_greater_than_number_of_leaves() {
        let mut tree = MerkleTree::new();

        let data = &TEST_DATA[0..5]; // 5 leaves
        for datum in data.iter() {
            tree.push(datum);
        }

        let proof = tree.prove(10);
        assert!(proof.is_none());
    }

    #[test]
    fn prove_returns_the_merkle_root_and_empty_proof_set_for_1_leaf() {
        let mut tree = MerkleTree::new();

        let data = &TEST_DATA[0..1]; // 1 leaf
        for datum in data.iter() {
            tree.push(datum);
        }

        let leaf_0 = leaf_sum(data[0]);

        {
            let proof = tree.prove(0).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, leaf_0);
            assert_eq!(set, ProofSet::default());
        }
    }

    #[test]
    fn prove_returns_the_merkle_root_and_proof_set_for_7_leaves() {
        let mut tree = MerkleTree::new();

        let data = &TEST_DATA[0..7]; // 7 leaves
        for datum in data.iter() {
            tree.push(datum);
        }

        //               07
        //              /  \
        //             /    \
        //            /      \
        //           /        \
        //          /          \
        //         /            \
        //       03              11
        //      /  \            /  \
        //     /    \          /    \
        //   01      05      09      \
        //  /  \    /  \    /  \      \
        // 00  02  04  06  08  10     12
        // 00  01  02  03  04  05     06

        let leaf_0 = leaf_sum(data[0]);
        let leaf_1 = leaf_sum(data[1]);
        let leaf_2 = leaf_sum(data[2]);
        let leaf_3 = leaf_sum(data[3]);
        let leaf_4 = leaf_sum(data[4]);
        let leaf_5 = leaf_sum(data[5]);
        let leaf_6 = leaf_sum(data[6]);

        let node_1 = node_sum(&leaf_0, &leaf_1);
        let node_5 = node_sum(&leaf_2, &leaf_3);
        let node_3 = node_sum(&node_1, &node_5);
        let node_9 = node_sum(&leaf_4, &leaf_5);
        let node_11 = node_sum(&node_9, &leaf_6);
        let node_7 = node_sum(&node_3, &node_11);

        {
            let proof = tree.prove(0).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_1);
            assert_eq!(set[1], node_5);
            assert_eq!(set[2], node_11);
        }
        {
            let proof = tree.prove(1).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_0);
            assert_eq!(set[1], node_5);
            assert_eq!(set[2], node_11);
        }
        {
            let proof = tree.prove(2).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_3);
            assert_eq!(set[1], node_1);
            assert_eq!(set[2], node_11);
        }
        {
            let proof = tree.prove(3).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_2);
            assert_eq!(set[1], node_1);
            assert_eq!(set[2], node_11);
        }
        {
            let proof = tree.prove(4).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_5);
            assert_eq!(set[1], leaf_6);
            assert_eq!(set[2], node_3);
        }
        {
            let proof = tree.prove(5).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], leaf_4);
            assert_eq!(set[1], leaf_6);
            assert_eq!(set[2], node_3);
        }
        {
            let proof = tree.prove(6).unwrap();
            let root = proof.0;
            let set = proof.1;

            assert_eq!(root, node_7);
            assert_eq!(set[0], node_9);
            assert_eq!(set[1], node_3);
        }
    }
}
