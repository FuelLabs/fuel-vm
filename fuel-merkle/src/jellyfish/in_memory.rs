use fuel_storage::{
    StorageAsMut as _,
    StorageAsRef,
    StorageInspect,
    StorageMutate,
};

use crate::jellyfish::merkle_tree::JellyfishMerkleTreeStorage;

use crate::{
    common::{
        Bytes32,
        StorageMap,
    },
    sparse::MerkleTreeKey,
    storage::Mappable,
};

use alloc::borrow::Cow;

/// The table of the JellyFish Merkle Tree's nodes.
#[derive(Debug, Clone)]
pub struct NodesTable;

impl Mappable for NodesTable {
    type Key = Self::OwnedKey;
    type OwnedKey = jmt::storage::NodeKey;
    type OwnedValue = jmt::storage::Node;
    type Value = Self::OwnedValue;
}
/// The table of the Binary Merkle Tree's values.
#[derive(Debug, Clone)]
pub struct ValuesTable;

impl Mappable for ValuesTable {
    type Key = Self::OwnedKey;
    type OwnedKey = jmt::KeyHash;
    type OwnedValue = (jmt::Version, jmt::OwnedValue);
    type Value = Self::OwnedValue;
}

#[derive(Debug, Clone)]

/// Auxiliary table that stores the latest version of the tree.
pub struct LatestRootVersionTable;

impl Mappable for LatestRootVersionTable {
    type Key = Self::OwnedKey;
    type OwnedKey = ();
    type OwnedValue = jmt::Version;
    type Value = Self::OwnedValue;
}

#[derive(Debug, Default, Clone)]
pub struct Storage {
    pub nodes: StorageMap<NodesTable>,
    pub values: StorageMap<ValuesTable>,
    pub latest_root_version: Option<<LatestRootVersionTable as Mappable>::OwnedValue>,
}

impl StorageInspect<NodesTable> for Storage {
    type Error = core::convert::Infallible;

    fn get(
        &self,
        key: &<NodesTable as Mappable>::Key,
    ) -> Result<Option<std::borrow::Cow<<NodesTable as Mappable>::OwnedValue>>, Self::Error>
    {
        self.nodes.storage::<NodesTable>().get(key)
    }

    fn contains_key(
        &self,
        key: &<NodesTable as Mappable>::Key,
    ) -> Result<bool, Self::Error> {
        self.nodes.storage::<NodesTable>().contains_key(key)
    }
}

impl StorageInspect<ValuesTable> for Storage {
    type Error = core::convert::Infallible;

    fn get(
        &self,
        key: &<ValuesTable as Mappable>::Key,
    ) -> Result<
        Option<std::borrow::Cow<<ValuesTable as Mappable>::OwnedValue>>,
        Self::Error,
    > {
        self.values.storage::<ValuesTable>().get(key)
    }

    fn contains_key(
        &self,
        key: &<ValuesTable as Mappable>::Key,
    ) -> Result<bool, Self::Error> {
        self.values.storage::<ValuesTable>().contains_key(key)
    }
}

impl StorageInspect<LatestRootVersionTable> for Storage {
    type Error = core::convert::Infallible;

    fn get(
        &self,
        _key: &<LatestRootVersionTable as Mappable>::Key,
    ) -> Result<
        Option<std::borrow::Cow<<LatestRootVersionTable as Mappable>::OwnedValue>>,
        Self::Error,
    > {
        let latest_root_version = &self.latest_root_version;
        if let Some(latest_root_version) = latest_root_version {
            Ok(Some(Cow::Borrowed(latest_root_version)))
        } else {
            Ok(None)
        }
    }

    fn contains_key(
        &self,
        _key: &<LatestRootVersionTable as Mappable>::Key,
    ) -> Result<bool, Self::Error> {
        Ok(self.latest_root_version.is_some())
    }
}

impl StorageMutate<NodesTable> for Storage {
    fn replace(
        &mut self,
        key: &<NodesTable as Mappable>::Key,
        value: &<NodesTable as Mappable>::Value,
    ) -> Result<Option<<NodesTable as Mappable>::OwnedValue>, Self::Error> {
        self.nodes.storage_as_mut().replace(key, value)
    }

    fn take(
        &mut self,
        key: &<NodesTable as Mappable>::Key,
    ) -> Result<Option<<NodesTable as Mappable>::OwnedValue>, Self::Error> {
        self.nodes.storage_as_mut().take(key)
    }
}

impl StorageMutate<ValuesTable> for Storage {
    fn replace(
        &mut self,
        key: &<ValuesTable as Mappable>::Key,
        value: &<ValuesTable as Mappable>::Value,
    ) -> Result<Option<<ValuesTable as Mappable>::OwnedValue>, Self::Error> {
        self.values.storage_as_mut().replace(key, value)
    }

    fn take(
        &mut self,
        key: &<ValuesTable as Mappable>::Key,
    ) -> Result<Option<<ValuesTable as Mappable>::OwnedValue>, Self::Error> {
        self.values.storage_as_mut().take(key)
    }
}

impl StorageMutate<LatestRootVersionTable> for Storage {
    fn replace(
        &mut self,
        _key: &<LatestRootVersionTable as Mappable>::Key,
        value: &<LatestRootVersionTable as Mappable>::Value,
    ) -> Result<Option<<LatestRootVersionTable as Mappable>::OwnedValue>, Self::Error>
    {
        let old_value = self.latest_root_version.take();
        self.latest_root_version = Some(value.clone());
        Ok(old_value)
    }

    fn take(
        &mut self,
        _key: &<LatestRootVersionTable as Mappable>::Key,
    ) -> Result<Option<<LatestRootVersionTable as Mappable>::OwnedValue>, Self::Error>
    {
        Ok(self.latest_root_version.take())
    }
}

#[derive(Clone)]
pub struct MerkleTree {
    storage: JellyfishMerkleTreeStorage<
        NodesTable,
        ValuesTable,
        LatestRootVersionTable,
        Storage,
    >,
}

impl MerkleTree {
    pub fn new() -> anyhow::Result<Self> {
        let storage = Storage::default();
        Ok(Self {
            storage: JellyfishMerkleTreeStorage::new(storage)?,
        })
    }

    /// Build a sparse Merkle tree from a set of key-value pairs. This is
    /// equivalent to creating an empty sparse Merkle tree and sequentially
    /// calling [update](Self::update) for each key-value pair. This constructor
    /// is more performant than calling individual sequential updates and is the
    /// preferred approach when the key-values are known upfront. Leaves can be
    /// appended to the returned tree using `update` to further accumulate leaf
    /// data.
    pub fn from_set<I, D>(set: I) -> Self
    where
        I: Iterator<Item = (MerkleTreeKey, D)>,
        D: AsRef<[u8]>,
    {
        let storage = Storage::default();
        let storage = JellyfishMerkleTreeStorage::from_set(storage, set)
            .expect("`Storage` can't return error");
        Self { storage }
    }

    /// Calculate the sparse Merkle root from a set of key-value pairs. This is
    /// similar to constructing a new tree from a set of key-value pairs using
    /// [from_set](Self::from_set), except this method returns only the root; it
    /// does not write to storage nor return a sparse Merkle tree instance. It
    /// is equivalent to calling `from_set(..)`, followed by `root()`, but does
    /// not incur the overhead of storage writes. This can be helpful when we
    /// know all the key-values in the set upfront and we will not need to
    /// update the set in the future.
    pub fn root_from_set<I, D>(set: I) -> anyhow::Result<Bytes32>
    where
        I: Iterator<Item = (MerkleTreeKey, D)>,
        D: AsRef<[u8]>,
    {
        let tree = Self::from_set(set);
        tree.root()
    }

    pub fn nodes_from_set<I, D>(
        set: I,
    ) -> anyhow::Result<(Bytes32, Vec<(jmt::storage::NibblePath, jmt::storage::Node)>)>
    where
        I: Iterator<Item = (MerkleTreeKey, D)>,
        D: AsRef<[u8]>,
    {
        let tree = Self::from_set(set);
        let root = tree.root()?;
        let storage_read_guard = tree.storage.storage.read();
        let nodes = storage_read_guard
            .nodes
            .inner()
            .iter()
            .map(|(k, v)| (k.nibble_path().clone(), v.clone()))
            .collect();
        Ok((root, nodes))
    }

    pub fn update(&mut self, key: MerkleTreeKey, data: &[u8]) {
        let _ = self.storage.update(key, data);
    }

    pub fn delete(&mut self, key: MerkleTreeKey) {
        let _ = self.storage.delete(key);
    }

    pub fn root(&self) -> anyhow::Result<Bytes32> {
        self.storage.root()
    }
}

#[cfg(test)]
mod test {
    use crate::{
        jellyfish::merkle_tree::EMPTY_ROOT,
        sparse::MerkleTreeKey,
    };
    use sha2::Sha256;

    use super::*;

    #[test]
    fn root_returns_the_empty_root_for_0_leaves() {
        let tree = MerkleTree::new().unwrap();

        let root = tree.root().unwrap();
        assert_eq!(root, EMPTY_ROOT);
    }

    #[test]
    fn adding_key_value_pair_works() {
        let mut tree = MerkleTree::new().unwrap();
        let initial_storage_version =
            tree.storage.storage.read().latest_root_version.unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        let storage = tree.storage.storage.read();
        let nodes = storage.nodes.inner();
        let values = storage.values.inner();
        // The version has been updated:
        assert_eq!(
            storage.latest_root_version.unwrap(),
            initial_storage_version + 1
        );
        // The root has been updated:
        assert_ne!(tree.root().unwrap(), EMPTY_ROOT);
        // There is exactly one value in the tree
        assert_eq!(values.len(), 1);
        let leaves = nodes
            .iter()
            .filter_map(|(_node_key, node)| match node {
                jmt::storage::Node::Leaf(leaf_node) => Some(leaf_node),
                _ => None,
            })
            .collect::<Vec<_>>();
        assert_eq!(leaves.len(), 1);
        let leaf_node = leaves[0];
        // The only node is a leaf

        let value = values.iter().next().unwrap();
        let (value_key_hash, (_version, preimage)) = value;
        // The key_hash of the leaf node is the same as the key_hash of the value
        assert_eq!(value_key_hash, &leaf_node.key_hash());
        let preimage_value_hash = jmt::ValueHash::with::<Sha256>(preimage);
        let expected_leaf_node =
            jmt::storage::LeafNode::new(value_key_hash.clone(), preimage_value_hash);
        assert_eq!(leaf_node, &expected_leaf_node);
    }

    #[test]
    fn adding_and_removing_key_value_pair_gives_the_empty_root() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        tree.delete(merkle_tree_key);
        let first_root = tree.root().unwrap();
        assert_eq!(first_root, EMPTY_ROOT);
        tree.update(merkle_tree_key, data);
        tree.delete(merkle_tree_key);
        let second_root = tree.root().unwrap();
        assert_eq!(second_root, EMPTY_ROOT);
    }

    #[test]
    fn updating_key_with_same_value_does_not_change_root() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        let first_root = tree.root().unwrap();
        tree.update(merkle_tree_key, data);
        let second_root = tree.root().unwrap();
        assert_eq!(first_root, second_root);
    }

    #[test]
    fn updating_same_key_changes_root() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data1 = b"data1";
        let data2 = b"data2";
        tree.update(merkle_tree_key, data1);
        let first_root = tree.root().unwrap();
        tree.update(merkle_tree_key, data2);
        let second_root = tree.root().unwrap();
        assert_ne!(first_root, second_root);
    }

    #[test]
    fn verify_exclusion_proof_for_empty_tree_succeeds() {
        let tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let proof = tree.storage.generate_proof(&merkle_tree_key).unwrap();
        assert!(proof.is_exclusion_proof());
        assert!(proof.verify(EMPTY_ROOT));
    }

    #[test]
    fn verify_inclusion_proof_succeeds() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        let root = tree.root().unwrap();
        let proof = tree.storage.generate_proof(&merkle_tree_key).unwrap();
        assert!(proof.is_inclusion_proof());
        assert!(proof.verify(root));
    }

    #[test]
    fn verify_inclusion_on_empty_tree_fails() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        let proof = tree.storage.generate_proof(&merkle_tree_key).unwrap();
        assert!(proof.is_inclusion_proof());

        let empty_tree = MerkleTree::new().unwrap();
        let empty_root = empty_tree.root().unwrap();
        assert!(!proof.verify(empty_root));
    }

    #[test]
    fn verify_wrong_value_for_key_fails() {
        let mut tree = MerkleTree::new().unwrap();
        let raw_key = b"key";
        let merkle_tree_key = MerkleTreeKey::new(raw_key);
        let data = b"data";
        tree.update(merkle_tree_key, data);
        let root = tree.root().unwrap();
        let proof = tree.storage.generate_proof(&merkle_tree_key).unwrap();
        assert!(proof.is_inclusion_proof());
        let tampered_proof = match proof {
            crate::jellyfish::merkle_tree::MerkleProof::Inclusion(inclusion_proof) => {
                crate::jellyfish::merkle_tree::MerkleProof::Inclusion(
                    crate::jellyfish::merkle_tree::InclusionProof {
                        value: b"Wrong value".to_vec(),
                        ..inclusion_proof
                    },
                )
            }
            crate::jellyfish::merkle_tree::MerkleProof::Exclusion(_exclusion_proof) => {
                panic!("It's an inclusion proof")
            }
        };
        assert!(!tampered_proof.verify(root));
    }

    #[test]
    fn verify_multiple_updates() {
        let num_updates = 1000;
        let mut tree = MerkleTree::new().unwrap();
        let updates: Vec<_> = (0..num_updates)
            .map(|i| {
                let raw_key = format!("key{}", i);
                let raw_key = raw_key.as_bytes();
                let merkle_tree_key = MerkleTreeKey::new(raw_key);

                let data = format!("data{}", i);
                (merkle_tree_key, data)
            })
            .collect();
        for (key, value) in updates.iter() {
            tree.update(*key, value.as_bytes());
        }
        let root = tree.root().unwrap();
        for (key, _value) in updates {
            let proof = tree.storage.generate_proof(&key).unwrap();
            assert!(proof.is_inclusion_proof());
            assert!(proof.verify(root));
        }
        // Generate proof for several non-existing nodes
        let non_existing_keys: Vec<_> = (0..num_updates)
            .map(|i| {
                let raw_key = format!("non-existing-key{}", i);
                let raw_key = raw_key.as_bytes();
                MerkleTreeKey::new(raw_key)
            })
            .collect();
        for key in non_existing_keys {
            let proof = tree.storage.generate_proof(&key).unwrap();
            assert!(proof.is_exclusion_proof());
            assert!(proof.verify(root));
        }
    }
}
