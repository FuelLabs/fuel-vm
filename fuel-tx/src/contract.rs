use crate::{CheckError, StorageSlot, Transaction};

use fuel_crypto::Hasher;
use fuel_merkle::binary::in_memory::MerkleTree as BinaryMerkleTree;
use fuel_merkle::sparse::in_memory::MerkleTree as SparseMerkleTree;
use fuel_types::{Bytes32, Bytes8, ContractId, Salt};

use alloc::vec::Vec;
use core::iter;

#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
/// Deployable representation of a contract code.
pub struct Contract(Vec<u8>);

impl Contract {
    /// Calculate the code root of the contract, using [`Self::root_from_code`].
    pub fn root(&self) -> Bytes32 {
        Self::root_from_code(self)
    }

    /// Calculate the code root from a contract.
    ///
    /// <https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/identifiers.md#contract-id>
    pub fn root_from_code<B>(bytes: B) -> Bytes32
    where
        B: AsRef<[u8]>,
    {
        let mut tree = BinaryMerkleTree::new();

        bytes
            .as_ref()
            .chunks(Bytes8::LEN)
            .map(|c| {
                if c.len() == Bytes8::LEN {
                    // Safety: checked len chunk
                    unsafe { Bytes8::from_slice_unchecked(c) }
                } else {
                    // Potential collision with non-padded input. Consider adding an extra leaf
                    // for padding?
                    let mut b = [0u8; 8];

                    let l = c.len();
                    b[..l].copy_from_slice(c);

                    b.into()
                }
            })
            .for_each(|l| tree.push(l.as_ref()));

        tree.root().into()
    }

    /// Calculate the root of the initial storage slots for this contract
    pub fn initial_state_root<'a, I>(storage_slots: I) -> Bytes32
    where
        I: Iterator<Item = &'a StorageSlot>,
    {
        let mut tree = SparseMerkleTree::new();

        storage_slots.for_each(|s| tree.update(s.key(), s.value().as_ref()));

        tree.root().into()
    }

    /// The default state root value without any entries
    pub fn default_state_root() -> Bytes32 {
        Self::initial_state_root(iter::empty())
    }

    /// Calculate and return the contract id, provided a salt, code root and state root.
    ///
    /// <https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/identifiers.md#contract-id>
    pub fn id(&self, salt: &Salt, root: &Bytes32, state_root: &Bytes32) -> ContractId {
        let mut hasher = Hasher::default();

        hasher.input(ContractId::SEED);
        hasher.input(salt);
        hasher.input(root);
        hasher.input(state_root);

        ContractId::from(*hasher.digest())
    }
}

impl From<Vec<u8>> for Contract {
    fn from(c: Vec<u8>) -> Self {
        Self(c)
    }
}

impl From<&[u8]> for Contract {
    fn from(c: &[u8]) -> Self {
        Self(c.into())
    }
}

impl From<&mut [u8]> for Contract {
    fn from(c: &mut [u8]) -> Self {
        Self(c.into())
    }
}

impl From<Contract> for Vec<u8> {
    fn from(c: Contract) -> Vec<u8> {
        c.0
    }
}

impl AsRef<[u8]> for Contract {
    fn as_ref(&self) -> &[u8] {
        self.0.as_ref()
    }
}

impl AsMut<[u8]> for Contract {
    fn as_mut(&mut self) -> &mut [u8] {
        self.0.as_mut()
    }
}

impl TryFrom<&Transaction> for Contract {
    type Error = CheckError;

    fn try_from(tx: &Transaction) -> Result<Self, Self::Error> {
        match tx {
            Transaction::Create(create) => TryFrom::try_from(create),
            _ => Err(CheckError::TransactionScriptOutputContractCreated { index: 0 }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use fuel_types::{bytes::WORD_SIZE, Bytes64};
    use itertools::Itertools;
    use quickcheck_macros::quickcheck;
    use rand::{rngs::StdRng, RngCore, SeedableRng};
    use rstest::rstest;

    // safe-guard against breaking changes to the code root calculation for valid
    // sizes of bytecode (multiples of instruction size in bytes (half-word))
    #[rstest]
    fn code_root_snapshot(#[values(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 100)] instructions: usize) {
        let mut rng = StdRng::seed_from_u64(100);
        let code_len = instructions * WORD_SIZE / 2;
        let mut code = alloc::vec![0u8; code_len];
        rng.fill_bytes(code.as_mut_slice());

        // compute root
        let root = Contract::root_from_code(code);

        // take root snapshot
        insta::with_settings!(
            {snapshot_suffix => format!("instructions-{}", instructions)},
            {
                insta::assert_debug_snapshot!(root);
            }
        );
    }

    // validate code_root is always equivalent to contract.root
    #[quickcheck]
    fn contract_root_matches_code_root(instructions: u8) -> bool {
        let mut rng = StdRng::seed_from_u64(100);
        let code_len = instructions as usize * WORD_SIZE / 2;
        let mut code = alloc::vec![0u8; code_len];
        rng.fill_bytes(code.as_mut_slice());
        let contract = Contract::from(code.clone());
        // compute root
        let code_root = Contract::root_from_code(code);
        let contract_root = contract.root();
        code_root == contract_root
    }

    #[rstest]
    fn state_root_snapshot(
        #[values(Vec::new(), vec![Bytes64::new([1u8; 64])])] state_slot_bytes: Vec<Bytes64>,
    ) {
        let slots: Vec<StorageSlot> = state_slot_bytes.iter().map(Into::into).collect_vec();
        let state_root = Contract::initial_state_root(&mut slots.iter());
        // take root snapshot
        insta::with_settings!(
            {snapshot_suffix => format!("state-root-{}", slots.len())},
            {
                insta::assert_debug_snapshot!(state_root);
            }
        );
    }

    #[test]
    fn default_state_root_snapshot() {
        let default_root = Contract::default_state_root();
        insta::assert_debug_snapshot!(default_root);
    }
}
