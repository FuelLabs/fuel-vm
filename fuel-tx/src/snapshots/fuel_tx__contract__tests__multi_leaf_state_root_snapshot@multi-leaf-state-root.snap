---
source: fuel-tx/src/contract.rs
expression: root
---
0x573bf26ecbfae1dfe079dbcea08defc35e280787b0be24a4683c0e3a48cee837
